### PR TITLE
Use out-of-proc MSBuild evaluation

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Fixes https://github.com/dotnet/tye/issues/594

I have tested to ensure that the original issue where installing a 5.0 SDK would break apps has been fixed with this change. With this change, supporting new TFMs wouldn't require tye itself to be re-targeted.

So far testing the perf with frontend-backend sample gives:

Before:
backend.csproj
Evaluation Took: 1500.9242ms
frontend.csproj
Evaluation Took: 463.2118ms

After:
backend.csproj
Evaluation Took: 1733.3304ms
frontend.csproj
Evaluation Took: 1766.8971ms

So there is a performance penalty with running out of proc, especially with additional projects.

### CODE REVIEW

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net5.0</TargetFramework>
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:4]: Why downgrade these?

**Review by jkotalik — APPROVED** [review]
Change itself looks good, @glennc @davidfowl I think this is a price we'll need to pay for now to support .NET 5.

**Comment by davidfowl** [comment]
I'm not sure I want to pay this price. I'd like to discuss different approaches, like possibly doing more computation of all projects in a single pass rather than trigger an invocation per project.

**Review by davidfowl — CHANGES_REQUESTED** [review]
Lets try to see if we can do this in a bigger batch.

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj (hunk 1 above)
Comment (JunTaoLuo) [reply src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:4]: The TFMs were upgraded to support "net5.0" projects. Now that's no longer needed so I've reverted them to netcoreapp3.1 which is LTS.

**Comment by JunTaoLuo** [comment]
Updated numbers:

After:
backend.csproj and frontend.csproj: 2207.3488ms

I'll collect some numbers with more projects tomorrow.

File: src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Core/Microsoft.Tye.Core.csproj:4]: But 5.0 is cool and new.

File: src/Microsoft.Tye.Core/ApplicationFactory.cs [hunk 2]
```diff
@@ -74,6 +77,87 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
...
+                    var msbuildEvaluationResult = await ProcessUtil.RunAsync(
+                        "dotnet",
+                        $"build " +
+                            $"\"{projectPath}\" " +
+                            $"/p:CustomAfterMicrosoftCommonTargets={Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "ProjectEvaluation.targets")} " +
+                            $"/nologo",
+                        throwOnError: false,
+                        workingDirectory: directory.DirectoryPath);
+
+                    // If the build fails, we're not really blocked from doing our work.
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/ApplicationFactory.cs:132]: What can cause the build to fail here? Would the failure to evaluate a project be shown when we throw a CommandException?

**Review by jkotalik — APPROVED** [review]

File: src/Microsoft.Tye.Core/ApplicationFactory.cs (hunk 2 above)
Comment (JunTaoLuo) [reply src/Microsoft.Tye.Core/ApplicationFactory.cs:132]: I didn't run into any cases where this fails but to be defensive, I've added the details of the output to the debug logs.

**Pull request merged**